## The statistical model of FINEMAP

> Benner et al., 2016. Bioinformatics. 
> FINEMAP: efficient variable selection using summary data from genome-wide association studies

## simulate genotype and phenotypes 

1. Make 5 variants in 3 LD blocks and 2 causal variants
2. Phenotype follows: $$y = \sum_{c \in C}\beta_c*g_c + N(\epsilon|0, \sigma^2I)$$

```{r}
g1 = rbinom(1e3, 2, .5)
noise = which(rbinom(1e3, 1, .05) == 1)
g2 = g1; g2[noise] = 2 - g2[noise]

g3 = rbinom(1e3, 2, .2)
noise = which(rbinom(1e3, 1, .05) == 1)
g4 = g3; g4[noise] = 2 - g4[noise]

g5 = rbinom(1e3, 2, .3)

g = cbind(g1, g2, g3, g4, g5)
g = apply(g, 2, scale)

library(lattice); levelplot(cor(g))

gc = g[, c(2, 4)]
fc = colMeans(gc) / 2
bc = c(1, 2)

e = rnorm(1e3, 0, 1)
y = gc %*% bc + e

var(y)
sum(diag(var(gc)) * bc^2) + var(e)

f1 = apply(g, 2, function(g1) summary(lm(y ~ g1)))
f1s = do.call(rbind, lapply(f1, function(x) coef(x)["g1", ]))
zh = f1s[, "t value"]

mr = t(g) %*% g / 1e3
(lh1 = 1 / sqrt(1e3) * sd(e) * solve(mr) %*% zh)

f2 = lm(y ~ g)
f2s = summary(f2)
(lh2 = f2s$coefficients[-1, "Estimate"]) # true model

data = data.frame(lh1, lh2)
data$c = c(0, 1, 0, 1, 0)
ggplot(data, aes(x = lh1, y = lh2, color = as.factor(c))) + geom_point() + xlab("Estimated") + ylab("True")
plot(lh1, lh2, xlab = "Estimated", ylab = "True"); abline(0, 1)
```

## The formula below to estimate causal effect using summary statistics is not ideal

The non-zero estimation in the non-causal loci, sometime large values, makes it impossible to estimate the causal ones.
? With this problem, some tools still use this formula for causal inferences, such as GCTA, PAINTOR, and CAVIARBF.
? What are they doing?

$$\hat{\lambda} = n^{-1/2} * \sigma * R^{-1} * \hat{z}$$

## Priors

>> To make more accurate estimates, by specifying priors under the Bayesian modeling framework.

$$p(\lambda|\gamma) = N(\lambda|0, s_{\lambda}^2 \sigma^2 \Delta_{\gamma})$$
$\Delta_{\gamma}$ is a diagnal matrix with $\gamma$ on the diagnal. 
This is equivalent to specify that the prior of causal effect size for the zero-locations of $\gamma$ is $N(0, 0)$.
Further, this is a huge prior that guarantees the posterior estimates all equals to the prior mean: 0.

```{r}
rnorm(1e2, 0, 0)

library(rstan)

mystan = "

data {
  int N; // size
  vector[N] x; // predictor 
  vector[N] y; // outcome vector
}
parameters {
  real beta; // effect
  real<lower=0> sigma; // variance
}
model {
  beta ~ normal(0, 1); // 
//beta ~ normal(0, 1e-4); // this is to mimic sigma = 0
  y ~ normal(x * beta, sigma);  // likelihood
}

"
model = stan_model(model_code = mystan)

xx = rnorm(1e2, 0, 1)
yy = 2 * xx + rnorm(1e2, 0, 1)

data = list(N = 100, x = xx, y = yy)
myfit = sampling(model, data = data, chains = 2, iter = 2e2)
print(myfit)

z4 = do.call(cbind, z4)

```
